presto-hive/src/main/java/com/facebook/presto/hive/statistics/MetastoreHiveStatisticsProvider.java

```diff
@@ -85,441 +94,708 @@
-                            .filter(stats -> stats.containsKey(columnName))
-                            .map(stats -> stats.get(columnName))
-                            .map(stats -> getMinMaxAsPrestoNativeValues(stats, prestoType, timeZone))
-                            .collect(toImmutableList());
-
-                    // TODO[lo] Maybe we do not want to expose high/low value if it is based on too small fraction of
-                    //          partitions. And return unknown if most of the partitions we are working with do not have
-                    //          statistics computed.
-                    lowValueCandidates = ranges.stream()
-                            .filter(range -> range.getMin().isPresent())
-                            .map(range -> range.getMin().get())
-                            .collect(toImmutableList());
-                    highValueCandidates = ranges.stream()
-                            .filter(range -> range.getMax().isPresent())
-                            .map(range -> range.getMax().get())
-                            .collect(toImmutableList());
-                }
-                dataSize = calculateDataSize(statisticsSample, columnName, rowCount);
+        List<HivePartition> partitionsSample = getPartitionsSample(partitions, sampleSize);
+        try {
+            Map<String, PartitionStatistics> statisticsSample = statisticsProvider.getPartitionsStatistics(table, partitionsSample);
+            validatePartitionStatistics(table, statisticsSample);
+            return getTableStatistics(columns, columnTypes, partitions, statisticsSample);
+        }
+        catch (PrestoException e) {
+            if (e.getErrorCode().equals(HIVE_CORRUPTED_COLUMN_STATISTICS.toErrorCode()) && isIgnoreCorruptedStatistics(session)) {
+                log.error(e);
+                return TableStatistics.empty();
             }
+            throw e;
+        }
+    }
 
-            Comparator<Object> comparator = (leftValue, rightValue) -> {
-                Block leftBlock = nativeValueToBlock(prestoType, leftValue);
-                Block rightBlock = nativeValueToBlock(prestoType, rightValue);
-                return prestoType.compareTo(leftBlock, 0, rightBlock, 0);
-            };
-            Optional<Object> min = lowValueCandidates.stream().min(comparator);
-            Optional<Object> max = highValueCandidates.stream().max(comparator);
-            if (min.isPresent() && max.isPresent()) {
-                columnStatistics.setRange(createPrestoRange(prestoType, min.get(), max.get()));
+    @VisibleForTesting
+    static List<HivePartition> getPartitionsSample(List<HivePartition> partitions, int sampleSize)
+    {
+        checkArgument(sampleSize > 0, "sampleSize is expected to be greater than zero");
+
+        if (partitions.size() <= sampleSize) {
+            return partitions;
+        }
+
+        List<HivePartition> result = new ArrayList<>();
```

Comment:
Any reason not to use `ImmutableList.Builder`?

Reply:
The typical reason to use `ArrayList` is that immutable collections can't contain null. Can min/max be null?

Reply:
It has to be `List`, so we can check it for `contains`. I think if it was a `Set` things would be a little bit more clear, but because we have to return ordered `List` i decided to reuse the result for `contains` checks.

Reply:
LinkedHashSet?  (And then convert into a List upon returning)

Reply:
At the moment of checking for `contains`, the result includes only 2 elements. So in a worth case scenario there will be one additional equals to be made. But a single equals is still faster (or similar) than taking a hash and doing a hash lookup.

Reply:
Yes! You're right. I didn't check the code and thought you're checking contains in a loop. Sorry for that